jvm-agent/src/main/org/jetbrains/lincheck/jvm/agent/analysis/controlflow/BasicBlockControlFlowGraph.kt

```diff
@@ -0,0 +1,483 @@
+        for (e in normalExits) {
+            val last = lastOpcodeIndexOf(e.source) ?: continue
+            val lastInstruction = instructions.get(last)
+            require(isRecognizedIfJumpOpcode(lastInstruction.opcode)) {
+                """
+                    Normal loop exit fall-through edge must be produced by an IF* opcode at the end of the source block: 
+                    source=B${e.source}, target=B${e.target}, opcode=${lastInstruction.opcode}
+                """.trimIndent()
+            }
+         }
+    }
+
+    /**
+     * Compute dominator sets for each basic block using the classical iterative algorithm.
+     * The entry block is assumed to be block 0 when the graph is non-empty.
+     * Returns an array where index i holds the set of dominators of block i (including i).
+     *
+     * A node d dominates node n if every path from the entry node to n must go through d.
+     * By definition, every node dominates itself. The entry node dominates all nodes in
+     * a reducible control flow graph.
+     */
+    private fun computeDominators(): Array<Set<BasicBlockIndex>> {
+        val n = basicBlocks.size
+        if (n == 0) return emptyArray()
+
+        // Build predecessor lists; include only normal (non-exception) predecessors for loop analysis dominators.
+        // Dominators for loop/back-edge detection are typically based on normal control-flow.
+        val preds: Array<MutableSet<BasicBlockIndex>> = Array(n) { mutableSetOf<BasicBlockIndex>() }
+        for (e in edges) {
+            // Note: we count exception edges as well
+            val u = e.source
+            val v = e.target
+            if (u in 0 until n && v in 0 until n) {
+                preds[v].add(u)
+            }
+        }
+
+        val all: Set<BasicBlockIndex> = (0 until n).toSet()
+        val doms: Array<Set<BasicBlockIndex>> = Array(n) { all.toMutableSet() }
+        // Entry is block 0 when present
+        doms[0] = setOf(0)
+
+        var changed = true
+        while (changed) {
+            changed = false
+            for (b in 1 until n) {
+                val newSet = buildSet {
+                    // Intersection of dominators of predecessors: dom(b) = {b} + intersect(dom(p1), dom(p2), ...)
+                    val it = preds[b].iterator()
+                    addAll(doms[it.next()])
+                    while (it.hasNext()) {
+                        retainAll(doms[it.next()])
+                    }
+                    add(b)
+                }
+                if (newSet != doms[b]) {
+                    doms[b] = newSet
+                    changed = true
+                }
+            }
+        }
+        return Array(n) { doms[it].toSet() }
+    }
+
+    /**
+     * Compute loops using dominators and back-edge detection.
+     * A back edge is an edge u -> h (non-exception) where h dominates u.
+     * For each header h, the loop body is the union of natural loops of its back edges.
```

Comment:
What are "natural loops"? Do we mention and explain this term somewhere in our docs?